### ISSUE TITLE
upper quantile normalization code

### ISSUE THREAD

**Issue by jingchunzhu** [ucscXena/python-scripts#1]
input: 
xena genomicMatrix file -supplied on the command line
gene file (each row is a gene) - supplied on the command line

output: 
upper quantile normalized genomicMatrix file, 
the offset parameter for each sample

1. download a gene expression xena genomicMatrix on a TCGA data, such as http://ec2-52-23-185-93.compute-1.amazonaws.com/datapages/?dataset=TCGA.COAD.sampleMap/HiSeqV2&host=https://tcga.xenahubs.net
sample ids are column header, gene names are on the first column'

2. using the unit in the url to revert back the value to before the log transformation, e.g. "log2(norm_count +1)" -> revert back to "norm_count" , where 1 is the pseudo count

3. Using the reverted values, for each column in the file downloaded in 1,  the script will only rank the genes belong to the gene file, identify the upper (top) 75% ranked value (100% is the gene with the highest value),  assign this as the upper quantile value for this column, then divide every value in the column by the upper quantile value, multiply by 1 million, plus the pseudo count value (e.g. 1 in the example), then log2 transform

log2(x+1) =>  x  => log2(x/uq * 1E6 +1) 

4. a good gene list genes that are expression in at least 90% TCGA samples and 90% GTEx samples: is https://github.com/ucscXena/python-scripts/blob/master/geneLists/GTEX_TCGA_genes

5. output the upper quantile normalized matrix

6. output the upper quantile value for each column in a separate file in this format
column id   upper quantile value
column id   upper quantile value

7. optional: include an optional  probeMap file for the situation the download genomicMatrix file does not use gene names. The probeMap file can be downloaded, containing mapping information between gene and first column in the matrix.

**Comment by GiriB** [comment]
@jingchunzhu  Is there a possibility that genes present in the `gene` file are not there in the `genomicMatrix` file ? If yes, what should be done in such a case ?

**Comment by jingchunzhu** [comment]
@GiriB yes and very common. I am in the middle of this one. I think the script is pretty much done now. I am working on evaluation of the results the scripts generated to see if the three upper quantile normalizations make the data better (i.e. less batch effect). 

My first draft of the code is here: https://github.com/jingchunzhu/cgDataNew/tree/master/normalization

Currently it calls xenaAPI in another directory to get data directly from the hub. Ideally xenaAPI needs to be a standalone python package. I am not sure how to do that yet.

**Comment by GiriB** [comment]
@jingchunzhu  I also had started with this. Just completed [first draft too.](https://github.com/GiriB/python-scripts/blob/master/upper_quantile_normalize.py ) Can you take a look ?  
I'll take a look at your code too. I was asked by @maryjgoldman to work with this issue (and also try to dockerise the script).  
Do let me know if there is anything else I can help with :)

**Comment by jingchunzhu** [comment]
@GiriB  sorry and thank you for your code. I need the script for our presentation in a week, so I have to get this done quickly. I will use your code to compare results against my own.

I will create another issue to put xena python API into a python "package" or "module", so people can do something like "pip install xenaAPI" to get the python connection.

**Comment by jingchunzhu** [comment]
@GiriB see https://github.com/ucscXena/python-scripts/issues/2

**Comment by GiriB** [comment]
@jingchunzhu Thanks. I'll start working on #2.

**Comment by jingchunzhu** [comment]
@GiriB thank you so much for the Xena Python package.

**Comment by GiriB** [comment]
@jingchunzhu I'm glad that I could be of help! :)
